### ISSUE TITLE
whereReactedBy fails when reacter/reactant are in different database than the Laravel Love tables

### ISSUE THREAD

**Issue by vesper8** [cybercog/laravel-love#163]
I have a project that uses 6 different databases, most of all my important data is in a database different from the main Laravel database that I called "core". However my main User database is in the main "core" database, almost everything else is in a 'data' database.

Using this setup I am able to use Laravel Love without a hitch so far.. I can add/remove reactions and all the tables and counts are working correctly.

My main use of Laravel Love is that users can like other users. Very simple.

Today I tried to do this:

```
$users = User::with('profile')->whereReactedBy($user, 'Like')->get();
```

And I ran into this error:

```
<!--
Illuminate\Database\QueryException: SQLSTATE[42S02]: Base table or view not found: 1146 Table 'core.love_reactants' doesn't exist (SQL: select * from `core`.`users` where exists (select * from `love_reactants` where `core`.`users`.`love_reactant_id` = `love_reactants`.`id` and exists (select * from `love_reactions` where `love_reactants`.`id` = `love_reactions`.`reactant_id` and `reacter_id` = 6176 and `reaction_type_id` = 1)) and `core`.`users`.`deleted_at` is null) in file /Volumes/Memories/Photos/2012/PictureBright/DEV/_NS/2020/omega-api.nomadsoulmates.com/vendor/laravel/framework/src/Illuminate/Database/Connection.php on line 671
```

I've run into such issues with other relations cross-database and usually the fix is as simple as going inside the model file and prefixing the table with the database name

So something like this:

```
class Comment extends Model
{
    protected $table = 'data.matches';
```

The problem is I can't do this with the Love models since they are in your codebase. I haven't tried binding my own models in place of yours so I can add this one extra line.

I did define in the love.php configuration file that it should use my database connection and things have worked so far. I guess this would work if you also prefixed the database name to the definition of your tables in your vendor models when the user specifies a custom database connection.

If you have another solution to offer I would love to hear it.

Thank you very much!

**Comment by antonkomarev** [comment]
@vesper8 am I right that you've tried to modify table property in vendor file and it worked well?

```php
final class Reactant extends Model implements
    ReactantContract
{
    protected $table = 'data.love_reactants';
```

**Comment by vesper8** [comment]
I haven't tried that... because that was too dirty... I did however come up with this raw query that works around the problem

```
    public function scopeWhereFavoritedBy($query, $user)
    {
        $query->whereRaw(
            "id in (select id from `core`.`users` where exists (
                        select * from `data`.`love_reactants` 
                            where `core`.`users`.`love_reactant_id` = `data`.`love_reactants`.`id` 
                            and exists (
                                select * from `data`.`love_reactions` 
                                    where `data`.`love_reactants`.`id` = `data`.`love_reactions`.`reactant_id` 
                                    and `reacter_id` = ?
                                    and `reaction_type_id` = 1)))", 
                                [$user->love_reacter_id]
        );
    }
```

Let me try your solution real quick just to confirm that's the solution that would make it work

**Comment by antonkomarev** [comment]
Prefixing table name with `storage.database.connection` wouldn't help because it is connection name, not schema/database.

**Comment by vesper8** [comment]
yes, I can confirm that adding `data.` in front of each of the 6 occurences of `protected $table = '` works : )

**Comment by vesper8** [comment]
right, yes I wasn't suggesting to prefix with connection name, you would have had to do a little reverse work to get the database name from the config/database.php file using the connection name and then prefix the actual database name

**Comment by vesper8** [comment]
I've simplified my issue by using `data` as my database name but in reality, `data` is my connection name and my database name is `ns_api_data`

**Comment by antonkomarev** [comment]
Another one approach - allow to customize all models table names via config.

I've introduced such solution in another one package 4 years ago:
https://github.com/cmgmyr/laravel-messenger/blob/master/config/config.php

**Comment by vesper8** [comment]
That could definitely work and allow even greater flexibility. As a fix though, getting the database name from the custom connection and applying that to the table names would work well without breaking anything

**Comment by vesper8** [comment]
either or would be of great help so I don't have to rely on a dirty raw query ;-) or even worse.. appending the database names to the vendor files on deployment using sed ;p

**Comment by antonkomarev** [comment]
It is working, but I'm not sure it was the best design decision. I think we should think about it a bit more and don't rush to add. 

From the first sight your proposal looks easier to implement, but it wouldn't work for all the databases this way. In PostgreSQL there is a schema which should be taken into account. Then it might look like this:
```php
public function getTable(): string
{
    $connectionName = $this->getConnectionName();

    if ($connectionName === null) {
        return $this->table;
    }

    $tablePath = '';

    $databaseName = Config::get("database.connections.{$connectionName}.database");
    if ($databaseName !== null) {
        $tablePath .= $databaseName . '.';
    }

    $schemaName = Config::get("database.connections.{$connectionName}.schema");
    if ($schemaName !== null) {
        $tablePath .= $schemaName . '.';
    }

    return $tablePath . $this->table;
}
```

**Comment by antonkomarev** [comment]
Solution from previous comment looks like Laravel doesn't use `getConnection()` method while making queries, so we need to do this job manually. It looks as strange hacky solution. Maybe we need to fix it on the framework level.

**Comment by vesper8** [comment]
Hrmm.. I like your solution.. tough nut to crack about `getConnectionName()` not being used while making queries... perhaps this can provide an answer: https://stackoverflow.com/questions/49186818/laravel-dynamically-set-table-name-in-model

**Comment by antonkomarev** [comment]
@vesper8 what Laravel version are you using?

**Comment by vesper8** [comment]
I'm on 7.9.2 at the moment

**Comment by antonkomarev** [comment]
[This solution](https://github.com/cybercog/laravel-love/issues/163#issuecomment-621561310) is working, but only for fetching. Adding reactions via reacter facade has an issue:
```
Illuminate/Database/QueryException with message 'SQLSTATE[42000]: Syntax error or access violation: 1064 You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '.`love_reactions` (`rate`, `reaction_type_id`, `reactant_id`, `reacter_id`, `upd' at line 1 (SQL: insert into `sandbox-laravel-love-data`.`sandbox-laravel-love-data`.`love_reactions` (`rate`, `reaction_type_id`, `reactant_id`, `reacter_id`, `updated_at`, `created_at`) values (1, 1, 3, 1, 2020-04-30 21:50:05, 2020-04-30 21:50:05))'
```

**Comment by antonkomarev** [comment]
It seems that Laravel using database connection for insert records, but ignoring it for fetching, so I'm gonna dig into the framework sources to try to fix it.

**Comment by vesper8** [comment]
Hrmm.. I'm surprised you ran into this issue when adding reactions because adding reactions was already working for me despite my unusual scenario where my users are in a different database then the love tables. This is how I was toggling reactions on and off through this helper method I added on my User model

```
    public function toggleLike($model)
    {
        $userReacter = $this->viaLoveReacter();

        if ($userReacter->hasReactedTo($model, 'like')) {
            $userReacter->unreactTo($model, 'like');
        } else {
            $userReacter->reactTo($model, 'like');
        }
    }
```

My user model table definition is `protected $table = 'ns_api_core.users';` (with the database prefixed) so maybe that's why I was having no issues and you are now

**Comment by antonkomarev** [comment]
By the way, you could use low level API to achieve the same result.

```
$reacter = $user->getLoveReacter();

$reactantIds = Reaction::query()
  ->where('reacter_id', $reacter->getId())
  ->where('reaction_type_id', ReactionType::fromName('Like')->getId())
  ->get()
  ->pluck('reactant_id');

$users = User::whereIn('love_reactant_id', $reactantIds)->get();
```

It's not that good to use it because it's more fragile to changes between major versions, but it's working solution.

**Comment by antonkomarev** [comment]
> My user model table definition is `protected $table = 'ns_api_core.users';` (with the database prefixed) so maybe that's why I was having no issues and you are now

Exactly. Without explicitly defined database for User model it leads to an error.

**Comment by antonkomarev** [comment]
We've got an issue with query builder. This code:
```php
$sql = User::whereHas('loveReactant')->toSql();
dd($sql);
```

Generates SQL:
```sql
select * from `users` where exists (select * from `love_reactants` where `users`.`love_reactant_id` = `love_reactants`.`id`)
```

**Comment by vesper8** [comment]
I'm not sure this one is solvable... your solution above takes care of one
side of the problem, perhaps merge that and simply include a note that if
you want cross-database relationships to work then you have to prefix your
models with their respective database names (and all databases need to be
on the same server). This is just how Laravel works.

**Comment by vesper8** [comment]
Seems like https://github.com/hoyvoy/laravel-cross-database-subqueries/issues might provide some secret sauce to make this possible... but I think pulling that functionality in is overkill as this is quite an edge case.. I'd be very happy with just being able to set the database prefix on the Laravel Love models somehow, through the configuration file, and being responsible for adding the prefix to my own models myself

**Comment by antonkomarev** [comment]
This issue was raised many times before:
- https://github.com/laravel/framework/issues/4330
- https://github.com/laravel/framework/issues/4649
- https://github.com/laravel/framework/issues/17624
- https://github.com/laravel/framework/issues/22985
- https://github.com/laravel/framework/issues/24398

And there was an attempt to add it into the core:
- https://github.com/laravel/framework/pull/23036

I will think about ability to configure table names.

**Comment by antonkomarev** [comment]
@vesper8 I've created a new thread with implementation design #164

I'm going to describe pros/cons of each solution and try to find out more of them. If you have any ideas - feel free to write yours.

**Comment by antonkomarev** [comment]
Issue was fixed in [Laravel Love v8.4.0](https://github.com/cybercog/laravel-love/releases/tag/8.4.0).
